client/executor/src/native_executor.rs

```diff
@@ -269,12 +281,108 @@ impl<D: NativeExecutionDispatch> RuntimeInfo for NativeExecutor<D> {
 			runtime_code,
 			ext,
 			false,
-			|_instance, version, _ext|
+			|_module, _instance, version, _ext|
 				Ok(version.cloned().ok_or_else(|| Error::ApiError("Unknown version".into()))),
 		)
 	}
 }
 
+pub struct RuntimeInstanceSpawn {
+	module: Arc<dyn sc_executor_common::wasm_runtime::WasmModule>,
+	forks: parking_lot::Mutex<HashMap<u32, std::sync::mpsc::Receiver<Vec<u8>>>>,
+	counter: std::sync::atomic::AtomicU32,
+	scheduler: Box<dyn sp_core::traits::SpawnNamed>,
+}
+
+impl sp_io::RuntimeSpawn for RuntimeInstanceSpawn {
+	fn dyn_dispatch(&self, func: u32, data: Vec<u8>) -> u32 {
+		use codec::Encode as _;
+
+		let new_handle = self.counter.fetch_add(1, std::sync::atomic::Ordering::Relaxed);
+
+		let (sender, receiver) = std::sync::mpsc::channel();
+		self.forks.lock().insert(new_handle, receiver);
+
+		let module = self.module.clone();
+		let scheduler = self.scheduler.clone();
+		self.scheduler.spawn("forked runtime invoke", Box::pin(async move {
+			let module = AssertUnwindSafe(module);
+			let result = with_externalities_safe(
+				&mut new_async_externalities(scheduler, module.clone())
+					.expect("Failed to setup externalities for async context"),
+				move || {
+
+					// FIXME: Should be refactored to shared "instance factory".
+					// Instatiating wasm here every time is suboptimal at the moment, shared
+					// pool of istances should be used.
+					let instance = module.new_instance().expect("Failed to create new instance for fork");
+
+					let mut dispatch_data = Vec::new();
+					func.encode_to(&mut dispatch_data);
+					data.encode_to(&mut dispatch_data);
+					instance.call("dyn_dispatch", &dispatch_data[..]).expect("Failed to invoke instance.")
```

Comment:
This is actually a quite important change.

Basically, this function becomes a part of the Substrate Runtime Api, even though optional (i.e. if you want to use the forking/spawning)

I wonder though, what is the reason for going that? Wouldn't it be more clean to  refer to a table index directly? That would allow to avoid exporting `dyn_dispatch`. That would also allow to avoid the double wrapping of the payload

Reply:
Yeah, as discussed, would be cleaner, updated